Calib stage rotation

Baharis opened the pull request:
### Context

An interesting feature of Instamatic, a program designed primarily with rotation electron diffraction in mind, is the fact that none of its currently available experiments allows to set stage rotation speed in an explicit and straightforward way. The only frame designed with rotation speed expressed in comprehensible units, cRED Gatan, is not in use. The simplistic cRED FEI allows setting the rotation speed setting as accepted by FEI microscopes to any values between 0 and 1, and most of the other experiments rely on the Joel settings, which similarily permit one of the 12 possible setting speeds. Both of these have little meaning without additional calibration or context.

In this PR, I suggest adding a new calibration class, responsible for calculating and translating between arbitrary speed settings and concrete speed values. `CalibStageRotation` can be generated by running `instamatic.calibrate_stage_rotation`, serializing into a small `calib_stage_rotation.yaml`. The process is done by running a small stage rotation at fixed speed, measuring time needed for each point, and then fitting all experimental points to equation `time = (alpha_pace * alpha_span + alpha_windup) / speed + delay`. As such, the relation between stage rotation setting, span, speed, and needed time is described using only a few floating point variables:

- `alpha_pace` – time taken by the stage to move by one degree at speed setting 1;
- `alpha_windup` – a variable delay scaling with inverse of speed setting, e.g. due to backlash;
- `delay` – a constant delay not scaling with inverse of speed setting, e.g. due to communication;
- `speed_options` – optional list of the limitations on speed settings e.g. limits or set values;

Stage rotation calibration assumes that rotation speed grows inverse-proportionally with the speed setting. Calibration file produced for Joel or simulated microscope sand may look as follows:
```
alpha_pace: 1.1985327615474075
alpha_windup: -0.030993505932929235
delay: 0.1118754654744266
speed_options:
  options: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
```

By default, stage rotation calibration will have been produced by rotating the stage 120 points and fitting the resulting grid of 120 points (12 speed settings x 10 alpha distances) to the aforementioned equation. The entire process may take up to an hour and needs to be done only once, unless the goniometer is modified or otherwise reconfigured.

<img width="640" height="480" alt="image" src="https://github.com/user-attachments/assets/45cf43f6-816b-4d98-8791-8d6a68edbe00" />

After the process, the calibration class instance is available via `CalibStageRotation.from_file()`. It can be read and interpreted 1) manually, 2) via simple `span_speed_to_time`, `span_time_to_speed`, and `speed_time_to_span`, or 3) via `plan_rotation(target_pace)` method, which produces a `RotationPlan` with closest `pace` available, `speed` setting needed to achieve it, as well as the `total_delay` added to the total movement time given current settings. A cRED experiment at desired pace can be then conducted roughly as follows:
```python
rotation_plan = rotation_calibration.plan_rotation(frame_separation / oscillation_angle)
total_exposure = rotation_plan.pace * rotation_angle - rotation_plan.dead_time
with self.ctrl.beam.unblanked()
    movie = self.ctrl.get_movie(n_frames=n, exposure=total_exposure/n)
    self.ctrl.stage.set_with_speed(a=rotation_angle, speed=rotation_plan.speed, wait=False):
    for image, header in movie:
        images.append(image)
        headers.append(header)
```

This PR alongside #135 and many other is for the purpose of implementing a general cRED / FastEDT interface. Other changes introduced in this PR include mostly bug fixes and code maintenance issues, as listed below.

### Minor changes
- Add `CalibStageRotation`, which allows expressing rotation speed in sec/deg rather than abstract units.

### Bug fixes and code maintenance
- Improve formatting when requesting `instamatic.calibrate_movie_delays` to be run;
-

stefsmeets commented on src/instamatic/calibrate/calibrate_stage_rotation.py:
```diff
@@ -0,0 +1,395 @@
+    total_delay: float  # total goniometer delay: delay + alpha_windup / speed
+
+
+class CalibStageRotation:
```
This could also be a dataclass to get rid of the __init__

stefsmeets commented on src/instamatic/calibrate/calibrate_stage_rotation.py:
```diff
@@ -0,0 +1,395 @@
+        log(f'{self} saved to {path}.')
+
+
+class SpanSpeedTime(NamedTuple):
```
As a matter of fact, this could also be a dataclass 😅

Baharis replied:
Makes code less readable, see joint comment below.

stefsmeets commented on src/instamatic/calibrate/calibrate_stage_rotation.py:
```diff
@@ -0,0 +1,395 @@
+    time: float  # time taken to travel span with speed expressed in seconds
+
+
+def plot_stage_rotation(
```
Can this be a method on `CalibStageRotation`?

stefsmeets commented on src/instamatic/calibrate/calibrate_stage_rotation.py:
```diff
@@ -0,0 +1,395 @@
+            for speed in speed_range:
+                with ctrl.stage.rotation_speed(speed=float(speed)):
+                    ctrl.stage.a = 0.0
+                    for at, as_ in zip(alpha_targets, alpha_spans):
```
```suggestion
                    for target, span in zip(alpha_targets, alpha_spans):
```

stefsmeets commented on src/instamatic/calibrate/calibrate_stage_rotation.py:
```diff
@@ -0,0 +1,395 @@
+    starting_stage_speed = ctrl.stage.get_rotation_speed()
+    ctrl.cam.block()
+    try:
+        try:
+            ctrl.stage.set_rotation_speed(12)
+            assert ctrl.stage.get_rotation_speed() == 12
+        except AssertionError:  # or any other raised if speed can't be set
+            speed_range_default = np.linspace(0.01, 0.2, num=20)
+            speed_options = FEI_ROTATION_SPEED_OPTIONS
+        else:
+            speed_range_default = np.arange(1, 13, step=1)
+            speed_options = JEOL_ROTATION_SPEED_OPTIONS
```
This seems like a bit round-about way to determine if we are working with a JEOL or FEI machine. You can just query `ctrl.tem.name`:

```suggestion
        if ctrl.tem.name == 'fei':
            speed_range_default = np.linspace(0.01, 0.2, num=20)
            speed_options = FEI_ROTATION_SPEED_OPTIONS
		else:
            speed_range_default = np.arange(1, 13, step=1)
            speed_options = JEOL_ROTATION_SPEED_OPTIONS
```

Baharis replied:
I try not to rely on `ctrl.tem.name`, as it seems to cause issues, see joint comment below.

stefsmeets commented on src/instamatic/calibrate/calibrate_stage_rotation.py:
```diff
@@ -0,0 +1,395 @@
+                with ctrl.stage.rotation_speed(speed=float(speed)):
+                    ctrl.stage.a = 0.0
+                    for at, as_ in zip(alpha_targets, alpha_spans):
+                        with timer() as t:
+                            ctrl.stage.a = float(at)
+                        calib_points.append(SpanSpeedTime(as_, speed, t()))
```
Although the contextmanager for the timer is clever, I think this is way easier to read:

```suggestion
                        t1 = time.perf_counter()
                        ctrl.stage.a = float(at)
                        t2 = time.perf_counter()
                        calib_points.append(SpanSpeedTime(span, speed, t2-t1))
```

stefsmeets commented on src/instamatic/calibrate/calibrate_stage_rotation.py:
```diff
@@ -0,0 +1,395 @@
+    starting_stage_alpha = ctrl.stage.a
+    starting_stage_speed = ctrl.stage.get_rotation_speed()
+    ctrl.cam.block()
+    try:
+        try:
+            ctrl.stage.set_rotation_speed(12)
+            assert ctrl.stage.get_rotation_speed() == 12
+        except AssertionError:  # or any other raised if speed can't be set
+            speed_range_default = np.linspace(0.01, 0.2, num=20)
+            speed_options = FEI_ROTATION_SPEED_OPTIONS
+        else:
+            speed_range_default = np.arange(1, 13, step=1)
+            speed_options = JEOL_ROTATION_SPEED_OPTIONS
+        speed_range = speed_range or speed_range_default
+
+        total = len(speed_range) * len(alpha_spans)
+        log(f'Starting rotation speed calibration based on {total} points.')
+
+        with tqdm(total=total) as progress_bar:
+            for speed in speed_range:
+                with ctrl.stage.rotation_speed(speed=float(speed)):
+                    ctrl.stage.a = 0.0
+                    for at, as_ in zip(alpha_targets, alpha_spans):
+                        with timer() as t:
+                            ctrl.stage.a = float(at)
+                        calib_points.append(SpanSpeedTime(as_, speed, t()))
+                        progress_bar.update(1)
+    finally:
+        ctrl.stage.set(a=starting_stage_alpha)
+        ctrl.stage.set_rotation_speed(starting_stage_speed)
+        ctrl.cam.unblock()
```
I think this whole bit would be easier to read if it was flattened a bit.

Baharis replied:
Applied ~ a bit, see joint comment below.

stefsmeets commented on src/instamatic/calibrate/calibrate_stage_rotation.py:
```diff
@@ -0,0 +1,395 @@
+    log('model time   = (alpha_pace * alpha_span + alpha_windup) / speed + delay')
+
+    c = CalibStageRotation(*p[0], speed_options=speed_options)
+    c.to_file(None if outdir is None else Path(outdir) / CALIB_STAGE_ROTATION)
```
See comment on `c.to_file()`

```suggestion
    c.to_file(outdir=outdir)
```

Baharis replied:
Agreed and applied, see joint comment below.

stefsmeets commented on src/instamatic/calibrate/calibrate_stage_rotation.py:
```diff
@@ -0,0 +1,395 @@
+            k: float(v) if isinstance(v, float) else v.to_dict() for k, v in vars(self).items()
+        }
+
+    def to_file(self, path: Optional[str] = None) -> None:
```
Based on the code below (line 345), I think it makes more sense to ask for an outdir here:

```suggestion
    def to_file(self, outdir: Optional[str] = None) -> None:
```

Baharis replied:
Agreed and applied, see joint comment below.

Baharis commented:
@stefsmeets Thank you for the review. Initially I was quite against many of the suggestions, but ultimately I think it resulted in a code that is both shorter and cleaner.

I modified all of `FloatOptions`, `FloatOptionsLimited`, `FloatOptionsListed`, and `CalibStageRotation` to be dataclasses. I think I missed them initially because I needed a lot of added functionality, but the `dataclass` wrapper allows great flexibility. Also the fact that `asdict` works recursively removes a lot of complexity.

It's not all roses though. The `FloatOptions.__new__`, for example, should IMO stay: when unpacking the calibration from yaml, `CalibStageRotation.speed_options` needs to be initialized as `FloatOptionsLimited` (if yaml has "options") or `FloatOptionsListed` (if yaml has "lower_lim" and "upper_lim"). This could be done by moving the contents of `__new__` to `CalibStageRotation.from_file`, but then in every place where calibration is initialized, you'd need custom logic for selecting one of these classes again. It seems more convenient to have a custom constructor that automatically does this for you. Still, I tried making it less scary:
```python
@dataclass
class FloatOptions(ABC):
    """Store valid float options & help finding the nearest available one."""

    def __new__(cls, *args, **kwargs) -> Self:
        """Init subclass based on kwargs, as needed to save/load calib@yaml."""
        if cls is not FloatOptions:
            return super().__new__(cls)  # avoids circular __new__
        if 'lower_lim' in kwargs and 'upper_lim' in kwargs:
            return FloatOptionsLimited(**kwargs)
        elif 'options' in kwargs:
            return FloatOptionsListed(**kwargs)
        else:
            raise ValueError("Can't determine child class based on kwargs.")

    @abstractmethod
    def nearest(self, to: float) -> float:
        """Find and return the nearest available float option."""
```

The most benefits come from converting `FloatOptions`s into dataclasses. `CalibStageRotation` does benefit from conversion, but then requires a hefty `__post_init__` because all of its instances must be of specific type, as shown before. Also note a clean use of the `FloatOptions` constructor:
```
    def __post_init__(self) -> None:
        self.alpha_pace = float(self.alpha_pace)
        self.alpha_windup = float(self.alpha_windup)
        self.delay = float(self.delay)
        if isinstance(self.speed_options, dict):
            self.speed_options = FloatOptions(**self.speed_options)
```

As a matter of fact, `SpanSpeedTime` could be theoretically a dataclass as well, but, interestingly, it being a tuple makes the code shorter and cleaner! That is because `np.array(list[NamedTuple])` becomes a 2D array of floats, but `np.array(list[DataclassInstance])` – a 1D array of dataclasses.

As for the suggestion to use `if ctrl.tem.name == 'fei'` instead of checking if speed setting 12 is available, as stupid as it looks, I believe this is actually a more future-proof solution. Your example illustrates the reason: there is not one TEMs that uses `FEI_ROTATION_SPEED_OPTIONS` – there are two: `fei` and `fei_simu`. In fact, there should be more: I am using a Tecnai microscope, but because `tecnai` it is not a vaild Instamatic setting, I need to trick the program into thinking I'm using `fei_simu` and having two different settings on the camera PC and microscope PC. This over-reliance on names already caused me issues in #108, and keeps causing them as seen in #132. Here I use duck-typing is a way to protect against it.

As for the `timer` context, yea, it was probably excessive so I removed it – but I still think it's cool!

I agree with the style changes, simplifications, `to_file` modification, all applied. I flattened the `with tqdm` piece a bit by removing timer and renamed some variables to be more verbose, but I actually believe it is more readable than alternatives. Difference in preferences I guess, we can discuss it.

Hope this version is cleaner and more to your liking @stefsmeets !

Baharis commented on src/instamatic/calibrate/calibrate_stage_rotation.py:
```diff
@@ -0,0 +1,395 @@
+        return vars(self)
+
+
+class FloatOptionsLimited(FloatOptions):
+    def __init__(self, lower_lim: float, upper_lim: float) -> None:
+        self.lower_lim = lower_lim
+        self.upper_lim = upper_lim
+
+    def nearest(self, to: float) -> float:
+        return float(np.clip(to, self.lower_lim, self.upper_lim))
```
Agreed and applied, see joint comment below.

Baharis commented on src/instamatic/calibrate/calibrate_stage_rotation.py:
```diff
@@ -0,0 +1,395 @@
+        return float(np.clip(to, self.lower_lim, self.upper_lim))
+
+
+class FloatOptionsListed(FloatOptions):
+    def __init__(self, options: List[float]) -> None:
+        self.options = list(options)  # must be a list to be serializable
+
+    def nearest(self, to: float) -> float:
+        options = np.array(self.options)
+        return float(options[(np.abs(options - to)).argmin()])
```
Agreed and applied, see joint comment below.

Baharis commented on src/instamatic/calibrate/calibrate_stage_rotation.py:
```diff
@@ -0,0 +1,395 @@
+    print(s)
+
+
+class FloatOptions:
+    """Store valid float options & help finding the nearest available one."""
+
+    def __new__(cls, *args, **kwargs):
+        """Initialize one of subclasses based on the contents of kwargs."""
+        if cls is FloatOptions:
+            if 'lower_lim' in kwargs and 'upper_lim' in kwargs:
+                return FloatOptionsLimited(**kwargs)
+            elif 'options' in kwargs:
+                return FloatOptionsListed(**kwargs)
+            else:
+                raise ValueError("Can't determine child class based on kwargs.")
+        return super().__new__(cls)
+
+    @abc.abstractmethod
+    def __init__(self, **kwargs) -> None:
+        """All instance variables must be settable under the same names."""
+
+    @abc.abstractmethod
+    def nearest(self, to: float) -> float:
+        """Find and return the nearest available float option (L^1 norm)."""
+
+    @classmethod
+    def from_dict(cls, dict_: dict) -> Self:
+        """Initialize self from a dict to allow easy serialization."""
+        return cls(**dict_)
+
+    def to_dict(self) -> dict:
+        """Convert self to a dict to allow easy serialization."""
+        return vars(self)
```
Removing `__new__` would just move this logic elsewhere, and I think it's best kept here. See joint comment for further details.

stefsmeets commented:
Thanks @Baharis , I think you forgot to push the changes though!

> As for the suggestion to use `if ctrl.tem.name == 'fei'` instead of checking if speed setting 12 is available, as stupid as it looks, I believe this is actually a more future-proof solution. Your example illustrates the reason: there is not one TEMs that uses `FEI_ROTATION_SPEED_OPTIONS` – there are two: `fei` and `fei_simu`. In fact, there should be more: I am using a Tecnai microscope, but because `tecnai` it is not a vaild Instamatic setting, I need to trick the program into thinking I'm using `fei_simu` and having two different settings on the camera PC and microscope PC. This over-reliance on names already caused me issues in #108, and keeps causing them as seen in #132. Here I use duck-typing is a way to protect against it.

Yeah that makes sense. Perhaps it's an idea to add a new attribute to the tem class, something like `tem.manufacturer` to identify the api, and let `tem.name` be the model name or whatever the class self-identifies as.

> It's not all roses though. The `FloatOptions.__new__`, for example, should IMO stay: when unpacking the calibration from yaml, `CalibStageRotation.speed_options` needs to be initialized as either `FloatOptionsLimited` (if yaml has "options") or `FloatOptionsListed` (if yaml has "lower_lim" and "upper_lim"). This could be done by moving the contents of `__new__` to `CalibStageRotation.from_file`, but then in every place where calibration is initialized, you'd need custom logic for selecting one of these classes again. It seems more convenient to have a custom constructor that automatically does this for you.

Yup, dataclasses fall short there without parsing / validation. You could have a separate loader function that tries to choose the correct class. I have been using attrs and pydantic in newer projects to abstract this away, and automatically take care of this depending on the data types / data structure.

Baharis commented:
I further shortened the indented block in `calibrate_stage_rotation_live`, added `--mode` argument for a greater control over calibration without a need to manually modify files, and improved doc-strings. In particular, I added a note that running `instamatic.calibrate_stage_rotation -a "1,2,3" -s "8,10,12"` is great for a test / debug run and suffices for the simulated TEM.
